x-pack/legacy/plugins/lens/server/routes/index_stats.ts

```diff
@@ -0,0 +1,177 @@
+} from '../../../../../../src/legacy/server/index_patterns/service';
+
+type Document = Record<string, unknown>;
+
+type Fields = Array<{ name: string; type: string; esTypes?: string[] }>;
+
+export async function initStatsRoute(setup: CoreSetup) {
+  const router = setup.http.createRouter();
+  router.post(
+    {
+      path: '/index_stats/{indexPatternTitle}',
+      validate: {
+        params: schema.object({
+          indexPatternTitle: schema.string(),
+        }),
+        body: schema.object({
+          earliest: schema.string(),
+          latest: schema.string(),
+          timeZone: schema.maybe(schema.string()),
+          timeFieldName: schema.string(),
+          size: schema.number(),
+          fields: schema.arrayOf(
+            schema.object({
+              name: schema.string(),
+              type: schema.string(),
+            })
+          ),
+        }),
+      },
+    },
+    async (context, req, res) => {
+      const requestClient = context.core.elasticsearch.dataClient;
+
+      const indexPatternsService = new IndexPatternsService(requestClient.callAsCurrentUser);
+
+      const { earliest, latest, timeZone, timeFieldName, fields, size } = req.body;
+
+      try {
+        const indexPattern = await indexPatternsService.getFieldsForWildcard({
+          pattern: req.params.indexPatternTitle,
+          // TODO: Pull this from kibana advanced settings
+          metaFields: ['_source', '_id', '_type', '_index', '_score'],
+        });
+
+        const results = (await requestClient.callAsCurrentUser('search', {
+          index: req.params.indexPatternTitle,
+          body: {
+            query: {
+              bool: {
+                filter: [
+                  {
+                    range: {
+                      [timeFieldName]: {
+                        gte: earliest,
+                        lte: latest,
+                        time_zone: timeZone,
+                      },
+                    },
+                  },
+                ],
+              },
+            },
+            size,
+          },
+        })) as SearchResponse<Document>;
+
+        if (results.hits.hits.length) {
+          return res.ok({
+            body: recursiveFlatten(results.hits.hits, indexPattern, fields),
+          });
+        }
+        return res.ok({ body: {} });
+      } catch (e) {
+        if (e.isBoom) {
+          return res.internalError(e);
+        } else {
+          return res.internalError({
+            body: Boom.internal(e.message || e.name),
+          });
+        }
+      }
+    }
+  );
+}
+
+export function recursiveFlatten(
+  docs: Array<{
+    _source: Document;
+  }>,
+  indexPattern: FieldDescriptor[],
+  fields: Fields
+): Record<
+  string,
+  {
+    count: number;
+    cardinality: number;
+  }
+> {
+  const overallKeys: Record<
+    string,
+    {
+      count: number;
+      samples: unknown[];
+    }
+  > = {};
+
+  const expectedKeys: Record<string, boolean> = {};
```

Comment:
Should probably just use a set here: `const expectedKeys = new Set(fields.map(f => f.name));`

And on line 128: `if (!expectedKeys.has(field.name)) {` or a filter.

Reply:
Done